pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,460 @@
+        at the given point.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            args: arguments to the objective function
+            kwargs: keyword arguments to the objective function
+
+        Returns:
+            tuple[array[float], array[float]]: a tuple of NumPy arrays containing the gradient
+            :math:`\nabla f(x^{(t)})` and the variance of the gradient.
+        """
+        if isinstance(objective_fn, qml.ExpvalCost):
+            grads = self._single_shot_expval_gradients(objective_fn, args, kwargs)
+        elif isinstance(objective_fn, qml.tape.QNode) or hasattr(objective_fn, "device"):
+            grads = self._single_shot_qnode_gradients(objective_fn, args, kwargs)
+        else:
+            raise ValueError(
+                "The objective function must either be encoded as a single QNode or "
+                "an ExpvalCost object for the Rosalin optimizer. "
+            )
+
+        # grads will have dimension [max(self.s), *params.shape]
+        # For each parameter, we want to truncate the number of shots to self.s[idx],
+        # and take the mean and the variance.
+        gradients = []
+        gradient_variances = []
+
+        for i, grad in enumerate(grads):
+            p_ind = np.ndindex(*grad.shape[1:])
+
+            g = np.zeros_like(grad[0])
+            s = np.zeros_like(grad[0])
+
+            for idx in p_ind:
+                grad_slice = grad[(slice(0, self.s[i][idx]),) + idx]
+                g[idx] = np.mean(grad_slice)
+                s[idx] = np.var(grad_slice, ddof=1)
+
+            gradients.append(g)
+            gradient_variances.append(s)
+
+        return gradients, gradient_variances
+
+    def step(self, objective_fn, *args, **kwargs):
+        """Update trainable arguments with one step of the optimizer.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            *args : Variable length argument list for objective function
+            **kwargs : variable length of keyword arguments for the objective function
+
+        Returns:
+            list[array]: the new variable values :math:`x^{(t+1)}`.
+            If single arg is provided, list [array] is replaced by array.
+        """
+
+        self.trainable_args = set()
+
+        for index, arg in enumerate(args):
+            if getattr(arg, "requires_grad", True):
+                self.trainable_args |= {index}
+
+        if self.s is None:
+            # Number of shots per parameter
+            self.s = [
+                np.zeros_like(a, dtype=np.int64) + self.min_shots
+                for i, a in enumerate(args)
+                if i in self.trainable_args
+            ]
+
+        # keep track of the number of shots run
+        s = np.concatenate([i.flatten() for i in self.s])
+        self.max_shots = max(s)
+        self.shots_used = int(2 * np.sum(s))

```

Comment:
Why do we need to cast to `int` if `s` is of type `np.int64`?

Reply:
Very good question. At some point, there must be an operation that is implicitly casting `s` to float 🤔 I wasn't able to track it down